cishouseholds/derive.py

```diff
@@ -1807,34 +1807,45 @@ def assign_regex_match_result(
-        matches against the positive and negative regex patterns. Setting `debug_mode` to True will retain these columns
-        otherwise they are dropped.
+        (optional) the Spark-compatible regex pattern to NOT match against. If given and `column_name_to_assign` is not
+        None, then two additional columns of the form: f"{column_name_to_assign}_positive" &
+        f"{column_name_to_assign}_negative" are created which track the matches against the positive and negative regex
+        patterns respectively. Set `debug_mode` to True to expose these columns in the returned dataframe.
     column_name_to_assign
-        name of the output column which will contain the result of the RegEx pattern search
+        (optional) if this is none, then we return a PySpark Column object containing the result of the RegEx pattern
+        search, otherwise we return a DataFrame with `column_name_to_assign` as the column containing the result.
     debug_mode:
-        See `negative_regex_pattern` above.
+        Only relevant when `column_name_to_assign` is not None - See `negative_regex_pattern` above.
     """
+    positive_regex_match_result = any_column_matches_regex(columns_to_check_in, positive_regex_pattern)
+
     if negative_regex_pattern is None:
-        df = df.withColumn(column_name_to_assign, any_column_matches_regex(columns_to_check_in, positive_regex_pattern))
+        result = positive_regex_match_result
+        if column_name_to_assign is not None:
+            df = df.withColumn(column_name_to_assign, positive_regex_match_result)
     else:
-        df = (
-            df.withColumn(
-                f"{column_name_to_assign}_positive",
-                any_column_matches_regex(columns_to_check_in, positive_regex_pattern),
-            )
-            .withColumn(
-                f"{column_name_to_assign}_negative",
-                any_column_matches_regex(columns_to_check_in, negative_regex_pattern),
-            )
-            .withColumn(
-                column_name_to_assign,
-                F.col(f"{column_name_to_assign}_positive") & ~F.col(f"{column_name_to_assign}_negative"),
+        negative_regex_match_result = any_column_matches_regex(columns_to_check_in, negative_regex_pattern)
+        result = positive_regex_match_result & ~negative_regex_match_result
+        if column_name_to_assign is not None:
+            df = (
+                df.withColumn(
+                    f"{column_name_to_assign}_positive",
+                    positive_regex_match_result,
+                )
+                .withColumn(
+                    f"{column_name_to_assign}_negative",
+                    negative_regex_match_result,
+                )
+                .withColumn(
+                    column_name_to_assign,
+                    result,
+                )
             )
-        )
 
+    if column_name_to_assign is None:
```

Comment:
This early return could be moved above the `if` block above, to do the early return before the rest is intepreted. Then the `if column_name_to_assign is not None` block can be de-indented, as it will be applied to all other cases